### ISSUE TITLE
Get xmlwf and "make check" to support XML_UNICODE and XML_UNICODE_WCHAR_T with MinGW

### ISSUE THREAD

**Issue by hartwork** [libexpat/libexpat#33]
What we know:

* `XML_UNICODE_WCHAR_T` with `sizeof(wchar_t) == 4` is broken as you found; so it's trouble on Linux and more or less cosmetic on Windows
* `XML_UNICODE` with 2-byte units works to some extent, at least even on Linux (see demo patch attached)
* `XML_UNICODE_WCHAR_T` with `sizeof(wchar_t) == 2` has been in use on Windows (see Visual Studio project files)
* GCC has `-fshort-wchar` to turn `wchar_t` into unsigned short under the hood, but it requires libc to be built with that flag, else functions like `wprintf` are expecting 4-byte units while they are passed 2-byte ones.
```
# gcc -dM -E - < /dev/null | fgrep -i wchar
#define __WCHAR_MAX__ 2147483647
#define __WCHAR_MIN__ (-__WCHAR_MAX__ - 1)
#define __GCC_ATOMIC_WCHAR_T_LOCK_FREE 2
#define __WCHAR_TYPE__ int
#define __SIZEOF_WCHAR_T__ 4

# gcc -fshort-wchar -dM -E - < /dev/null | fgrep -i wchar
#define __WCHAR_MAX__ 65535
#define __WCHAR_MIN__ 0
#define __GCC_ATOMIC_WCHAR_T_LOCK_FREE 2
#define __WCHAR_TYPE__ short unsigned int
#define __SIZEOF_WCHAR_T__ 2
```
* If an application works with 2-byte units for most of its string processing, converting back-and-forth to/from 1-byte units means waste in space and runtime
* xmlwf does not (run or even) compile with `XML_UNICODE`, at least not on Linux
* If `XML_UNICODE` support is kept, we want that code to be covered.

What I think we should do:
* Keep `XML_UNICODE` support (because it's used and removal would be hard to justify with regard to waste of hardware resources)
* Throw compile errors for `__SIZEOF_WCHAR_T__ != 2` where detectable
* Make the test suite work with `XML_UNICODE` and without `XML_UNICODE_WCHAR_T` by using `L"..."` literals and converting them to unsigned-short based strings and back on the fly.  ~~So we can print them and cover `XML_UNICODE` even on Linux.  I consider the need to store and free strings the most annoying aspect, personally.~~
* Make xmlwf work for `XML_UNICODE` ~~everywhere~~ with MinGW for a start.

**Comment by RMJ10** [comment]
It being used is the kicker; if it wasn't used any more, I'd argue for chopping it out entirely as being the least dangerous approach.

I'll get on with xmlwf and the test suite work, then.  I may be some time... :-)

**Comment by RMJ10** [comment]
I've been poking around this for several days, so I should report what I've found here.

* `XML_UNICODE`/`XML_UNICODE_WCHAR_T` appears to be intended for Windows users to take advantage of the fact that Windows uses UTF-16 for its own resources.
* On Windows:
  * the library compiles for both `XML_UNICODE_WCHAR_T` and `!XML_UNICODE` cases without needing any changes.
  * `xmlwf` will compile and run with `XML_UNICODE_WCHAR_T` if you create a project for it.  No other work is needed.
  * The test suite needs a small amount of support to compile and run correctly with `XML_UNICODE_WCHAR_T`.  This falls into two categories:
    * Macros to hide whether the narrow (normal) or wide string functions and literals are being used, and
    * More vigorous policing of the difference between `char` and `XML_Char` in the tests.
* On Linux:
  * `XML_UNICODE_WCHAR_T` doesn't work (sizeof(wchar_t) == 4)
  * The library will compile with `XML_UNICODE` defined.
  * Both `xmlwf` and the test suite need painful levels of support to compile and run with `XML_UNICODE`.  In addition to the points above:
    * There is no string literal representation for 16-bit characters.  They can be written out as array initialisers, but the cost in readability is considerable.
    * An alternative approach would be to write a narrow string to 16-bit conversion function allocating fresh memory to each literal.  Additional functions would be needed to free the allocated memory.  This is the approach I took on my test branch; it is workable, but even using convenience macros it is not pretty and prone to leaking small amounts of memory.
    * There is no library support for 16-bit strings, so copy and comparison functions and the like would need to be written.

In short, `XML_UNICODE_WCHAR_T` on Windows is easy; `XML_UNICODE` on Linux is a bit horrific.

Unfortunately, not doing the work for Linux has consequences.  In particular, we cannot get full test coverage of the code on Linux without `XML_UNICODE`.  Being able to build it on Windows is no help; I haven't found a suitable free coverage tool that works with Visual Studio, and MinGW doesn't seem to come with gcov.

A possible alternative would be to write a conversion program to preprocess `xmlwf` and the test suite source files into the less readable form that would compile.  I'll have a little poke and see if I can put something together.

**Comment by RMJ10** [comment]
Unfortunately preprocessing string literals to unsigned short initializers doesn't help.  Initializers initialize arrays, literals are (effectively) arrays :-(  The only via approach for compiling on Linux with `XML_UNICODE` seems to be allocated literals as on my xmlwf-unicode branch.

**Comment by kwaclaw** [comment]
From README:
1. For UTF-16 output as unsigned short (and version/error
  strings as char), run:

  ./configure CPPFLAGS=-DXML_UNICODE

  For UTF-16 output as wchar_t (incl. version/error strings),
  run:

  ./configure CFLAGS="-g -O2 *-fshort-wchar"* \
  CPPFLAGS=-DXML_UNICODE_WCHAR_T


So maybe XML_UNICODE_WCHAR_T  works when compiled with -fshort-wchar?

**Comment by hartwork** [comment]
> So maybe XML_UNICODE_WCHAR_T works when compiled with -fshort-wchar?

To some extent but as long as libc was compiled without `-fshort-wchar`, our 2-byte `wchar_t`s are not printable with `wprintf` as that expects 4-byte `wchar_t`s on Linux.

**Comment by kwaclaw** [comment]
Why should we support UTF-16 on Unix anyway? It isn't native (unless there
are some exotic Unix variants where it is the default). Would it not be
more appropriate to simply make UTF-32 work on Unix?

Though - there is one possible use case I can think of: Anyone who wants to
access Expat from Java will need UTF-16 support, even on Unix.

Whats your opinion?

On Tue, Jun 13, 2017 at 5:42 PM, Sebastian Pipping <notifications@github.com

**Comment by RMJ10** [comment]
I'd like UTF-16 just to be able to finish the coverage tests.  The library works, but as I said it's converting the tests and xmlwf that are painful.

**Comment by hartwork** [comment]
> Would it not be more appropriate to simply make UTF-32 work on Unix?

Adding UTF-32 support for Linux would be ideal, but it's not trivial and will need quite a bit of new test cases if we don't want to regret its introduction later.

> Why should we support UTF-16 on Unix anyway?

It would be one way getting code coverage for that code. Not sure yet if the best :)

> Anyone who wants to access Expat from Java will need UTF-16 support, even on Unix.

Interesting, will keep that in mind.

**Comment by hartwork** [comment]
A quick update: The current plan is to get `make check` and xmlwf working for `./configure --host=i686-w64-mingw32 CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T'` so we get Unicode code covered using MinGW.

**Comment by hartwork** [comment]
The "make check" part was fixed by #132, xmlwf is still left to do.

**Comment by hartwork** [comment]
xmlwf was fixed by #144 and related additional commits. Closing.
